README.md
# sscan -- Sean's scanner

## Synopsis

sscan is a UI for scanning multi-page documents using the scanimage(1)
command and a flatbed scanner.

I wrote sscan because I often need to scan multi-page documents on my
GNU/Linux system, and I have only a flatbed scanner, without an ADF.

sscan was originally a crude Python script.  Now it is a somewhat more
robust Haskell program, relying on the [Brick][] library.

[Brick]: https://github.com/jtdaugherty/brick/

## Screenshot

![screenshot](screenshot.png)

## Installation

1. Ensure that your scanner has [SANE support][].
2. Install prerequisite utilities:
````
apt-get install ocrmypdf pdftk sane-utils haskell-stack imagemagick
````
3. Use [stack][] to build and install sscan:
````
stack install sscan
````

Note that stack will automatically download and install the various
Haskell dependencies of sscan -- unfortunately, these are not yet
likely to all be installable from your distribution's mirrors.

[SANE support]: http://www.sane-project.org/sane-supported-devices.html
[stack]: https://haskellstack.org/

## Configuration

sscan does not yet have a configuration file.  If it did, this could
be used to configure the available presets, the default settings on
startup, and the output directory.  Patches welcome.

## Usage

Ensure that `~/.local/bin` is in your PATH environment variable.

Then open a terminal, run `sscan` and follow the on-screen
instructions.  If you can't see all the key bindings, you will need to
increase the size of your terminal.

## Bugs/patches

By e-mail to `<spwhitton@spwhitton.name>`.

## License

Copyright (C) 2017  Sean Whitton

sscan is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your
option) any later version.

sscan is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see
[<http://www.gnu.org/licenses/>](http://www.gnu.org/licenses/).


NEWS.md
0.2 (2020-01-27)
----------------

- README tweaks.
- Switch CHANGELOG to NEWS.md.
- Stop adding document author metadata.
  Previously, the author's own initials were hardcoded into sscan.

0.1 (2017-03-12)
----------------

- Initial public release.


UI.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

module UI (runTheApp) where

import qualified Graphics.Vty           as V
import           Lens.Micro             ((&), (.~), (^.))

import           Brick.AttrMap
import           Brick.Main
import           Brick.Types
import           Brick.Widgets.Border   as B
import           Brick.Widgets.Center   as C
import           Brick.Widgets.Core

import           Brick.Widgets.DefnList
import           Presets
import           Types.Preset
import           Types.State

drawUI :: St -> [Widget ()]
drawUI st = [ui]
  where
    ui = vBox [ hBorderWithLabel (str "[ Status ]")
              , vLimit 3 $ C.center $ status
              , hBorderWithLabel (str "[ Current settings ]")
              , vLimit 8 $ C.center $ settingsBox
              , hBorderWithLabel (str "[ Presets ]")
              , vLimit (2 + length presets) $ C.center $ presetsBox
              , hBorderWithLabel (str "[ Actions ]")
              , vLimit 6 $ C.center $ actionsBox
              ]
    status = str $ maybe
        "Ready to scan first page"
        (\(ScanSess _ p _) -> "Scanned " ++ show p ++ " pages")
        (st^.stScanSess)
    settingsBox = defnList AlignRight Nothing
        [ ("run OCRmyPDF", if st^.stOCR then "yes" else "no")
        , ("colour data",   show $ st^.stColour)
        , ("page size",     show $ st^.stPaper)
        , ("DPI",           show $ st^.stDPI)
        , ("output format", show $ st^.stOutFormat)
        , ("output dir",    st^.stOutdir)
        ]
    presetsBox = defnList AlignLeft
        (Just $ V.withStyle V.currentAttr V.bold)
        (map (\(Preset k desc _) -> ([k], desc)) presets)
    actionsBox = defnList AlignLeft
        (Just $ V.withStyle V.currentAttr V.bold) $
        (if st^.stOutFormat == PDF
         then (ifScanSess st
                [ ("SPC", "scan next page")
                , ("RET", "scan final page")
                , ("q", "declare most recent scanned page was the final page")
                , ("ESC", "abort/restart scanning this document")
                ]
                [ ("SPC", "scan first page of multi-page document")
                , ("RET", "scan single page document")
                , ("q", "quit sscan")
                ]
              )
         else [ ("SPC|RET", "scan page")
              , ("q", "quit sscan")
              ]
        )

handleQ :: St -> EventM () (Next St)
handleQ st = halt $ ifScanSess st
    (setScanSessCommand Finalise st)
    (resetScanSess st)          -- misnomer: this causes sscan to exit

handleRET :: St -> EventM () (Next St)
handleRET st = halt $ setScanSessCommand FinalPage st

handleSPC :: St -> EventM () (Next St)
handleSPC st = halt $ setScanSessCommand
    (case st^.stOutFormat of
       PDF -> NextPage
       PNG -> FinalPage)
    st

handleESC :: St -> EventM () (Next St)
handleESC st = ifScanSess st
    (halt $ setScanSessCommand Abort st)
    (continue st)

handleHotKey :: St -> Char -> EventM () (Next St)
handleHotKey st 'q' = handleQ st
handleHotKey st ' ' = handleSPC st
handleHotKey st 'o' = continue $ updateSt st
    (\s -> s & stOCR .~ (not $ st^.stOCR))
handleHotKey st 'c' = continue $ updateSt st
    (\s -> s & stColour .~ (cycleColour $ st^.stColour))
handleHotKey st 'p' = continue $ updateSt st
    (\s -> s & stPaper .~ (cyclePaper $ st^.stPaper))
handleHotKey st c = continue $ updateSt st $
    case lookupPreset c of
      Just (Preset _ _ f) -> f
      _                   -> id

appEvent :: St -> BrickEvent () e -> EventM () (Next St)
appEvent st (VtyEvent e) =
    case e of
      V.EvKey (V.KEnter) []  -> handleRET st
      V.EvKey (V.KEsc) []    -> handleESC st
      V.EvKey (V.KChar c) [] -> handleHotKey st c
      _                      -> continue st
appEvent st _ = continue st

theApp :: App St e ()
theApp =
    App { appDraw = drawUI
        , appChooseCursor = neverShowCursor
        , appHandleEvent = appEvent
        , appStartEvent = return
        , appAttrMap = const $ attrMap V.defAttr []
        }

runTheApp :: St -> IO St
runTheApp = defaultMain theApp


Types.hs
{-# LANGUAGE TemplateHaskell #-}

module Types where

import           Lens.Micro.TH        (makeLenses)

-- | Whether to do colour, grey or b&w scans
data Colour = Lineart | Greyscale | Colour
    deriving (Eq, Show)

cycleColour :: Colour -> Colour
cycleColour Lineart   = Greyscale
cycleColour Greyscale = Colour
cycleColour Colour    = Lineart

-- | Paper size to scan (determines both scanning area and PDF page
-- size)
data Paper = A4 | Letter | Photo | Auto
    deriving (Eq, Show)

cyclePaper :: Paper -> Paper
cyclePaper A4     = Letter
cyclePaper Letter = Photo
cyclePaper Photo  = Auto
cyclePaper Auto   = A4

-- | DPI to scan
type DPI = Int

data St =
    St { _stScanningSession :: Maybe FilePath -- ^ if a session is in
                                              -- progress, accmulate
                                              -- scans in this dir
       , _stOCR             :: Bool -- ^ whether to use OCRmyPDF
       , _stColour          :: Colour
       , _stPaper           :: Paper
       , _stDPI             :: DPI
       , _stOutdir          :: FilePath -- ^ where to save final PDFs
       }

makeLenses ''St


Presets.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

{-# LANGUAGE OverloadedStrings #-}

module Presets (presets, lookupPreset) where

import           Data.List    (find)
import           Lens.Micro   ((^.))

import           Types.Preset
import           Types.State

presets :: [Preset]
presets = [ Preset 'H' "handwritten notes" handwrittenNotes
          , Preset 'T' "typewritten docs" typewrittenDocs
          , Preset 'C' "standard DPI colour PDF, full page" colourPDF
          , Preset 'I' "standard DPI colour PNG, full page" colourPNG
          , Preset 'A' "standard DPI colour PNG, autocropped" colourCroppedPNG
          , Preset 'B' "black and white PDF, full page" bnwPDF
          , Preset 'P' "high DPI 6x4 photo" photo
          ]

lookupPreset :: Char -> Maybe Preset
lookupPreset c = find (\(Preset k _ _) -> c == k) presets

handwrittenNotes :: St -> St
handwrittenNotes = \st -> st
    { _stOCR     = False
    , _stColour  = Greyscale
    , _stDPI     = 75
    , _stPaper   = st^.stDefaultPaper
    , _stOutFormat = PDF
    }

typewrittenDocs :: St -> St
typewrittenDocs = \st -> st
    { _stOCR    = True
    , _stColour = Greyscale
    , _stDPI    = 300
    , _stPaper  = st^.stDefaultPaper
    , _stOutFormat = PDF
    }

colourPDF :: St -> St
colourPDF = \st -> st
    { _stOCR = False
    , _stColour = Colour
    , _stDPI = 300
    , _stPaper = st^.stDefaultPaper
    , _stOutFormat = PDF
    }

colourPNG :: St -> St
colourPNG = \st -> st
    { _stOCR = False
    , _stColour = Colour
    , _stDPI = 300
    , _stPaper = st^.stDefaultPaper
    , _stOutFormat = PNG
    }

colourCroppedPNG :: St -> St
colourCroppedPNG = \st -> st
    { _stOCR = False
    , _stColour = Colour
    , _stDPI = 300
    , _stPaper = Auto
    , _stOutFormat = PNG
    }

bnwPDF :: St -> St
bnwPDF = \st -> st
    { _stOCR = False
    , _stColour = Lineart
    , _stDPI = 150
    , _stPaper = st^.stDefaultPaper
    , _stOutFormat = PDF
    }

photo :: St -> St
photo = \st -> st
    { _stOCR = False
    , _stColour = Colour
    , _stDPI = 600
    , _stPaper = Photo
    , _stOutFormat = PNG
    }


Types/State.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

{-# LANGUAGE TemplateHaskell #-}

module Types.State where

import           Data.Maybe
import           Lens.Micro    ((&), (.~), (^.))
import           Lens.Micro.TH (makeLenses)

-- | Whether to do colour, grey or b&w scans
data Colour = Lineart | Greyscale | Colour
    deriving (Eq, Show)

cycleColour :: Colour -> Colour
cycleColour Lineart   = Greyscale
cycleColour Greyscale = Colour
cycleColour Colour    = Lineart

-- | Paper size to scan (determines both scanning area and PDF page
-- size)
data Paper = A4 | Letter | Photo | Auto
    deriving (Eq, Show)

cyclePaper :: Paper -> Paper
cyclePaper A4     = Letter
cyclePaper Letter = Photo
cyclePaper Photo  = Auto
cyclePaper Auto   = A4

-- | DPI to scan
type DPI = Int

-- | Output format
data OutputFormat = PDF | PNG
    deriving (Eq, Show)

-- | An active scanning session
data ScanSess = ScanSess
    -- | An action requested by the user, that has not yet been
    -- fulfilled
    Command
    -- | The number of pages scanned so far in this scanning session
    Int
    -- | Temporary directory containing pages scanned in this scanning
    -- session, if any have been scanned, and the session is multi-page
    (Maybe FilePath)

-- | An sscan operation selected by the user in the UI
data Command
    -- | Scan an additional page, or the first page, of a multi-page
    -- scanning session.
    = NextPage
    -- | Scan the final page of a multi-page scanning session, or the
    -- first page of a single-page scanning session
    | FinalPage
    -- | End a multi-page scanning session without scanning anymore
    -- pages, i.e., consider the most recently scanned page the final
    -- page
    | Finalise
    -- | Abandon the current scanning session.  Delete its tmpdir.
    | Abort

-- | Application state
data St =
    --
    St { _stScanSess     :: Maybe ScanSess -- ^ Nothing indicates user
                                           -- wants to quit sscan
       , _stOCR          :: Bool -- ^ whether to use OCRmyPDF
       , _stColour       :: Colour
       , _stPaper        :: Paper -- ^ currently selected paper size
       , _stDefaultPaper :: Paper -- ^ locale's default paper size
       , _stDPI          :: DPI
       , _stOutFormat    :: OutputFormat
       , _stOutdir       :: FilePath -- ^ where to save final PDFs
       }

makeLenses ''St

ifScanSess :: St -> a -> a -> a
ifScanSess st a b = if isJust $ st^.stScanSess then a else b

-- | Update a state when there is no scanning session in progress (the
-- state should not be changed when some pages have already been
-- scanned)
updateSt :: St -> (St -> St) -> St
updateSt st f = ifScanSess st st (f st)

resetScanSess :: St -> St
resetScanSess st = st & stScanSess .~ Nothing

setScanSessDir :: FilePath -> St -> St
setScanSessDir dir st = case st^.stScanSess of
  Nothing -> st
  Just (ScanSess c p _) ->
      st & stScanSess .~ (Just $ ScanSess c p (Just dir))

setScanSessCommand :: Command -> St -> St
setScanSessCommand c st = case st^.stScanSess of
  Just (ScanSess _ p d) ->
      st & stScanSess .~ (Just $ ScanSess c p d)
  Nothing ->
      st & stScanSess .~ (Just $ ScanSess c 0 Nothing)

incrementPages :: St -> St
incrementPages st = case st^.stScanSess of
  Nothing -> st
  Just (ScanSess c p d) ->
      st & stScanSess .~ (Just $ ScanSess c (p+1) d)

getLatestPage :: St -> Int
getLatestPage st = case st^.stScanSess of
  Nothing -> 0
  Just (ScanSess _ p _ ) -> p


Types/Preset.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

module Types.Preset where

import           Types.State

type PresetToggleKey = Char
type PresetDesc      = String
type PresetPreset    = St -> St

data Preset = Preset PresetToggleKey PresetDesc PresetPreset


Brick/Widgets/DefnList.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

{-# LANGUAGE OverloadedStrings #-}

module Brick.Widgets.DefnList (defnList, Align(..)) where

import           Data.Monoid
import qualified Data.Text          as T
import qualified Graphics.Vty       as V

import           Brick.Markup       (markup)
import           Brick.Types
import           Brick.Widgets.Core
import           Data.Text.Markup   ((@@))

data Align = AlignLeft | AlignRight
    deriving (Eq)

defnList :: Align -> Maybe V.Attr -> [(String, String)] -> Widget ()
defnList align attr defns = vBox $ line <$> defns
  where
    line (label, content) = markup $
        (T.pack label @@ labelAttr) <> (T.pack sep @@ V.defAttr)
        <> (T.pack (gap label content) @@ V.defAttr)
        <> (T.pack content @@ V.defAttr)

    gap x y
        | align == AlignRight =
              take (maxWidth - length x - length y - length sep) $
              repeat ' '
        | align == AlignLeft =
              take (maxLabelWidth - length x) $
              repeat ' '
    gap _ _ = " "

    maxWidth = maximum $
        map (\(x,y) -> length sep + length x + length y) defns
    maxLabelWidth = maximum $ map (\(x,_) -> length x) defns
    sep = if align == AlignRight then ":   " else ": "

    labelAttr = maybe V.defAttr id attr


Main.hs
{-

sscan --- text UI for scanning with SANE

Copyright (C) 2017  Sean Whitton

This file is part of sscan.

sscan is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

sscan is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with sscan.  If not, see <http://www.gnu.org/licenses/>.

-}

{-# LANGUAGE OverloadedStrings #-}

import           Control.Concurrent.Async (concurrently_)
import           Control.Monad            (unless, void, when)
import           Data.Time.Clock.POSIX    (getPOSIXTime, posixSecondsToUTCTime)
import           Data.Time.Format         (defaultTimeLocale, formatTime,
                                           iso8601DateFormat)
import           Lens.Micro               ((^.))
import           System.Directory         (getHomeDirectory, removeFile,
                                           renamePath, withCurrentDirectory)
import           System.Exit              (ExitCode (..))
import           System.FilePath          ((<.>), (</>))
import           System.IO                (IOMode (WriteMode), hClose,
                                           hGetContents, hPutStr, openFile,
                                           withBinaryFile, withFile)
import           System.IO.Temp           (withSystemTempDirectory)
import           System.Process

import           Types.State
import           UI

processScanSessDir :: St -> FilePath -> IO ()
processScanSessDir st dir = withCurrentDirectory dir $ do
    posix <- getPOSIXTime
    let stamp = show (round posix :: Int)
    logH <- openFile (logFile stamp) WriteMode -- TODO maybe AppendMode?
    void $ case st^.stOutFormat of
      PDF -> do
          -- 1. convert tiff->PDF
          void $ createProcessWait_ "convert"
              (proc "convert" (allPages ++ ["temp.pdf"]))
          -- 2. set metadata with pdftk
          renamePath (dir </> "temp.pdf") (dir </> "temp2.pdf")
          writeFile "metadata" (metadata posix)
          void $ createProcessWait_ "pdftk"
              (proc "pdftk"
                ["temp2.pdf", "update_info", "metadata", "output", "temp.pdf"])
              { std_in = NoStream
              , std_out = NoStream
              , std_err = UseHandle logH
              }
          -- 3. maybe ocrmypdf
          when (st^.stOCR) $ do
              renamePath (dir </> "temp.pdf") (dir </> "temp2.pdf")
              -- OCRmyPDF dies if stdout is not connected, so tell it
              -- to output to stdout
              withBinaryFile "temp.pdf" WriteMode $ \tempFile -> do
                (_, _, Just herr, p) <- createProcess_ "OCRmyPDF"
                -- we'd like to use --remove-background here, but then
                -- qpdf says that ocrmypdf's output is damaged.
                -- reported upstream
                  (proc "ocrmypdf" ["-c", "-i", "-r", "temp2.pdf", "-"])
                  { std_in = NoStream
                  , std_out = UseHandle tempFile
                  , std_err = CreatePipe
                  }
                -- OCRmyPDF does not yet have a --quiet option, so we
                -- emulate the chronic(1) utility
                exitCode <- waitForProcess p
                unless (exitCode == ExitSuccess) $
                  hGetContents herr >>= hPutStr logH
                hClose herr
          -- 4. qpdf (ocrmypdf invokes qpdf but it doesn't use
          -- --linearize, which shrinks the PDF, often substantially)
          void $ createProcessWait_ "qpdf"
              (proc "qpdf" ["--linearize", "temp.pdf", outFile stamp])
              { std_in = NoStream
              , std_out = NoStream
              , std_err = UseHandle logH
              }
      -- assume that only one page was scanned.  Not clear how we
      -- can avoid this assumption when producing a PNG
      PNG -> void $ createProcessWait_ "convert"
        (proc "convert" ["page1" <.> "tiff", outFile stamp])
          { std_in = NoStream
          , std_out = NoStream
          , std_err = UseHandle logH
          }
    hClose logH
    -- clean up the log file if it's empty
    finalLog <- readFile (logFile stamp)
    when (null finalLog) $ removeFile (logFile stamp)
  where
      logFile stamp = (st^.stOutdir) </> stamp <.> "log"
      outFile stamp = (st^.stOutdir) </> stamp <.> outExt
      outExt = case st^.stOutFormat of
        PDF -> "pdf"
        PNG -> "png"
      allPages = map (\n -> "page" ++ show n <.> "tiff") [1..(getLatestPage st)]
      metadata posix = let date = formatTime
                                  defaultTimeLocale
                                  (iso8601DateFormat Nothing)
                                  (posixSecondsToUTCTime posix)
                       in unlines [ "InfoBegin"
                                  , "InfoKey: Title"
                                  , "InfoValue: scan of " ++ date
                                  ]

makeInitialState :: IO St
makeInitialState = do
    home <- getHomeDirectory
    papersize <- init <$> readFile "/etc/papersize"
    let paper = if papersize == "letter" then Letter else A4
    return St
        { _stScanSess     = Nothing
        , _stOCR          = True
        , _stColour       = Greyscale
        , _stPaper        = paper
        , _stDefaultPaper = paper
        , _stDPI          = 300
        , _stOutFormat    = PDF
        , _stOutdir       = home </> "tmp"
        }

scanPage :: St -> FilePath -> IO Bool
scanPage st dir = withFile outF WriteMode $ \outH -> do
    exit <- createProcessWait_ "scanimage" (proc "scanimage" (scanimageArgs st))
        { std_in = NoStream
        , std_out = UseHandle outH
        , std_err = Inherit     -- let the user see progress bar
        }
    case exit of
      ExitSuccess -> return True
      ExitFailure c -> do
          putStrLn $ "scanimage exited with exit code " ++ show c ++ "!"
          putStrLn "This might because sscan tried to use an option that your"
          putStrLn "scanner does not support.  Please try another preset."
          putStrLn "Press any key to abort this scanning session..."
          void getChar
          return False
  where
      outF = dir </> "page" ++ (show $ getLatestPage st + 1) <.> "tiff"

scanimageArgs :: St -> [String]
scanimageArgs st =
    [ "-vp"
    , "--format=tiff"
    , "--resolution=" ++ show (st^.stDPI)
    , "--mode=" ++ case st^.stColour of
        Colour    -> "Color"
        Greyscale -> "Gray"
        Lineart   -> "Lineart"
    ] ++ case st^.stPaper of
           Auto -> ["--swcrop=yes"]
           _ -> [ "-x"
                , show $ case st^.stPaper of
                    A4     -> 210   :: Double
                    Letter -> 215.9 :: Double
                    Photo  -> 150   :: Double
                    Auto   -> 1     :: Double
                , "-y"
                , show $ case st^.stPaper of
                    A4     -> 297   :: Double
                    Letter -> 279.4 :: Double
                    Photo  -> 100   :: Double
                    Auto   -> 1     :: Double
                ]
-- other device-specific scanimage options the old script supported:
-- --swdespeck; --color-filter; --depth

-- TODO include usbreset.c as a cbit, and use FFI to optionally reset
-- the scanner's USB connection after a scan -- this is needed on some
-- scanners that I occasionally use

processCommand :: St -> IO ()
processCommand st = case st^.stScanSess of
  Nothing -> putStrLn "Waiting for OCR threads to complete..."
    >> return ()          -- quit sscan
  Just (ScanSess command _ maybeDir) -> case maybeDir of
    Nothing -> withSystemTempDirectory "sscan" $ \dir ->
      processCommand (setScanSessDir dir st)
    Just dir -> case command of
      Abort -> newSession
      NextPage -> scanPage st dir >>= \scanned ->
        if scanned
        then presentUI (incrementPages st)
        else newSession
      FinalPage -> scanPage st dir
                >> finaliseSession (incrementPages st) dir
      Finalise -> finaliseSession st dir
  where
    newSession = presentUI $ resetScanSess st
    finaliseSession st' dir =
        concurrently_ (processScanSessDir st' dir) newSession

presentUI    :: St -> IO ()
presentUI st = runTheApp st >>= processCommand

main :: IO ()
main = makeInitialState >>= presentUI

-- | Create a process, wait for it to finish, don't close any
-- handles used by the CreateProcess record
createProcessWait_ :: String -> CreateProcess -> IO ExitCode
createProcessWait_ s c = do
    (_, _, _, p) <- createProcess_ s c
    waitForProcess p
